### PR TITLE
`fragment.py`: refactoring, patches, and compatibility with RMS

### PR DESCRIPTION
**Pull request by JacksonBurns**
`fragment.py` should inherit from `molecule.py` rather than copy methods. This PR will attempt to implement this.

### CODE REVIEW

**Comment by JacksonBurns** [comment]
The latest output from `symilar` as well as the class diagrams for `fragment.py` and `molecule.py`:
classes

<details>
  <summary>Duplicated Lines:</summary>

32 similar lines in 2 files
==fragment.py:[265:302]
==molecule.py:[1018:1055]
       @property
       def fingerprint(self):
           """
           Fingerprint used to accelerate graph isomorphism comparisons with
           other molecules. The fingerprint is a short string containing a
           summary of selected information about the molecule. Two fingerprint
           strings matching is a necessary (but not sufficient) condition for
           the associated molecules to be isomorphic.

           Use an expanded molecular formula to also enable sorting.
           """
           if self._fingerprint is None:
               # Include these elements in this order at minimum
               element_dict = {'C': 0, 'H': 0, 'N': 0, 'O': 0, 'S': 0}
               all_elements = sorted(self.get_element_count().items(), key=lambda x: x[0])  # Sort alphabetically
               element_dict.update(all_elements)
               self._fingerprint = ''.join([f'{symbol}{num:0>2}' for symbol, num in element_dict.items()])
           return self._fingerprint

       @fingerprint.setter
       def fingerprint(self, fingerprint):
           self._fingerprint = fingerprint

       @property
       def inchi(self):
           """InChI string for this molecule. Read-only."""
           if self._inchi is None:
               self._inchi = self.to_inchi()
           return self._inchi

       @property
       def smiles(self):
           """SMILES string for this molecule. Read-only."""
           if self._smiles is None:
               self._smiles = self.to_smiles()
           return self._smiles


25 similar lines in 2 files
==fragment.py:[1500:1529]
==molecule.py:[1957:1986]
       def is_linear(self):
           """
           Return :data:`True` if the structure is linear and :data:`False`
           otherwise.
           """

           atom_count = len(self.vertices)

           # Monatomic molecules are definitely nonlinear
           if atom_count == 1:
               return False
           # Diatomic molecules are definitely linear
           elif atom_count == 2:
               return True
           # Cyclic molecules are definitely nonlinear
           elif self.is_cyclic():
               return False

           # True if all bonds are double bonds (e.g. O=C=O)
           all_double_bonds = True
           for atom1 in self.vertices:
               for bond in atom1.edges.values():
                   if not bond.is_double(): all_double_bonds = False
           if all_double_bonds: return True

           # True if alternating single-triple bonds (e.g. H-C#C-H)
           # This test requires explicit hydrogen atoms
           for atom in self.vertices:
               bonds = list(atom.edges.values())

16 similar lines in 2 files
==fragment.py:[1413:1433]
==molecule.py:[2594:2614]
           # Get a set of atom indices for each molecule
           atom_ids = set([atom.id for atom in self.atoms])
           other_ids = set([atom.id for atom in other.atoms])

           if atom_ids == other_ids:
               # If the two molecules have the same indices, then they might be identical
               # Sort the atoms by ID
               atom_list = sorted(self.atoms, key=lambda x: x.id)
               other_list = sorted(other.atoms, key=lambda x: x.id)

               # If matching atom indices gives a valid mapping, then the molecules are fully identical
               mapping = {}
               for atom1, atom2 in zip(atom_list, other_list):
                   mapping[atom1] = atom2

               return self.is_mapping_valid(other, mapping, equivalent=True, strict=strict)
           else:
               # The molecules don't have the same set of indices, so they are not identical
               return False


14 similar lines in 2 files
==fragment.py:[1331:1345]
==molecule.py:[2374:2388]
           except ValueError:
               logging.warning('Unable to check aromaticity by converting to RDKit Mol.')
           else:
               aromatic_rings = []
               aromatic_bonds = []
               for ring0 in rings:
                   aromatic_bonds_in_ring = []
                   # Figure out which atoms and bonds are aromatic and reassign appropriately:
                   for i, atom1 in enumerate(ring0):
                       if not atom1.is_carbon():
                           # all atoms in the ring must be carbon in RMG for our definition of aromatic
                           break
                       for atom2 in ring0[i + 1:]:
                           if self.has_bond(atom1, atom2):

14 similar lines in 2 files
==fragment.py:[1005:1020]
==molecule.py:[2037:2052]
       def count_internal_rotors(self):
           """
           Determine the number of internal rotors in the structure. Any single
           bond not in a cycle and between two atoms that also have other bonds
           is considered to be a pivot of an internal rotor.
           """
           count = 0
           for atom1 in self.vertices:
               for atom2, bond in atom1.edges.items():
                   if (self.vertices.index(atom1) < self.vertices.index(atom2) and
                           bond.is_single() and not self.is_bond_in_cycle(bond)):
                       if len(atom1.edges) > 1 and len(atom2.edges) > 1:
                           count += 1
           return count


13 similar lines in 2 files
==fragment.py:[1551:1565]
==molecule.py:[2238:2252]
               for i in range(atom.radical_electrons):
                   H = Atom('H', radical_electrons=0, lone_pairs=0, charge=0)
                   bond = Bond(atom, H, 1)
                   self.add_atom(H)
                   self.add_bond(bond)
                   if atom not in added:
                       added[atom] = []
                   added[atom].append([H, bond])
                   atom.decrement_radical()

           # Update the atom types of the saturated structure (not sure why
           # this is necessary, because saturating with H shouldn't be
           # changing atom types, but it doesn't hurt anything and is not
           # very expensive, so will do it anyway)

12 similar lines in 2 files
==fragment.py:[1333:1345]
==molecule.py:[2406:2418]
           else:
               aromatic_rings = []
               aromatic_bonds = []
               for ring0 in rings:
                   aromatic_bonds_in_ring = []
                   # Figure out which atoms and bonds are aromatic and reassign appropriately:
                   for i, atom1 in enumerate(ring0):
                       if not atom1.is_carbon():
                           # all atoms in the ring must be carbon in RMG for our definition of aromatic
                           break
                       for atom2 in ring0[i + 1:]:
                           if self.has_bond(atom1, atom2):

11 similar lines in 2 files
==fragment.py:[611:623]
==molecule.py:[1570:1582]
           for element, count in group.elementCount.items():
               if element not in element_count:
                   return False
               elif element_count[element] < count:
                   return False

           if generate_initial_map:
               keys = []
               atms = []
               initial_map = dict()
               for atom in self.atoms:
                   if atom.label and atom.label != '':

11 similar lines in 2 files
==fragment.py:[208:220]
==molecule.py:[1275:1287]
           # Copy connectivity values and sorting labels
           for i in range(len(self.vertices)):
               v1 = self.vertices[i]
               v2 = other.vertices[i]
               v2.connectivity1 = v1.connectivity1
               v2.connectivity2 = v1.connectivity2
               v2.connectivity3 = v1.connectivity3
               v2.sorting_label = v1.sorting_label
           other.multiplicity = self.multiplicity
           other.reactive = self.reactive
           return other


10 similar lines in 2 files
==fragment.py:[1295:1305]
==molecule.py:[1732:1742]
           # Check if multiplicity is possible
           n_rad = self.get_radical_count()
           multiplicity = self.multiplicity
           if not (n_rad + 1 == multiplicity or n_rad - 1 == multiplicity or
                   n_rad - 3 == multiplicity or n_rad - 5 == multiplicity):
               raise ValueError('Impossible multiplicity for molecule\n{0}\n multiplicity = {1} and number of '
                                'unpaired electrons = {2}'.format(self.to_adjacency_list(), multiplicity, n_rad))
           if raise_charge_exception:
               if self.get_net_charge() != 0:
                   raise ValueError('Non-neutral molecule encountered. '

9 similar lines in 2 files
==fragment.py:[371:381]
==molecule.py:[2212:2222]
       def get_charge_span(self):
           """
           Iterate through the atoms in the structure and calculate the charge span
           on the overall molecule.
           The charge span is a measure of the number of charge separations in a molecule.
           """
           abs_net_charge = abs(self.get_net_charge())
           sum_of_abs_charges = sum([abs(atom.charge) for atom in self.vertices])
           return (sum_of_abs_charges - abs_net_charge) / 2


9 similar lines in 2 files
==fragment.py:[175:184]
==molecule.py:[1665:1674]
       def draw(self, path):
           """
           Generate a pictorial representation of the chemical graph using the
           :mod:`draw` module. Use `path` to specify the file to save
           the generated image to; the image type is automatically determined by
           extension. Valid extensions are ``.png``, ``.svg``, ``.pdf``, and
           ``.ps``; of these, the first is a raster format and the remainder are
           vector formats.
           """

8 similar lines in 2 files
==fragment.py:[1571:1579]
==molecule.py:[2142:2150]
       def is_aryl_radical(self, aromatic_rings=None, save_order=False):
           """
           Return ``True`` if the molecule only contains aryl radicals,
           i.e., radical on an aromatic ring, or ``False`` otherwise.
           If no ``aromatic_rings`` provided, aromatic rings will be searched in-place,
           and this process may involve atom order change by default. Set ``save_order`` to
           ``True`` to force the atom order unchanged.
           """

8 similar lines in 2 files
==fragment.py:[1320:1330]
==molecule.py:[2363:2373]
           from rdkit.Chem.rdchem import BondType
           AROMATIC = BondType.AROMATIC

           if rings is None:
               rings = self.get_relevant_cycles()
               rings = [ring for ring in rings if len(ring) == 6]
           if not rings:
               return [], []

           try:

8 similar lines in 2 files
==fragment.py:[1308:1318]
==molecule.py:[2346:2357]
       def get_aromatic_rings(self, rings=None, save_order=False):
           """
           Returns all aromatic rings as a list of atoms and a list of bonds.

           Identifies rings using `Graph.get_smallest_set_of_smallest_rings()`, then uses RDKit to perceive aromaticity.
           RDKit uses an atom-based pi-electron counting algorithm to check aromaticity based on Huckel's Rule.
           Therefore, this method identifies "true" aromaticity, rather than simply the RMG bond type.

           The method currently restricts aromaticity to six-membered carbon-only rings. This is a limitation imposed
           by RMG, and not by RDKit.


8 similar lines in 2 files
==fragment.py:[576:585]
==molecule.py:[1490:1499]
           # Do the quick isomorphism comparison using the fingerprint
           # Two fingerprint strings matching is a necessary (but not
           # sufficient!) condition for the associated molecules to be isomorphic
           if self.fingerprint != other.fingerprint:
               return False
           # check multiplicity
           if self.multiplicity != other.multiplicity:
               return False


8 similar lines in 2 files
==fragment.py:[352:361]
==molecule.py:[1124:1133]
       def remove_bond(self, bond):
           """
           Remove the bond between atoms `atom1` and `atom2` from the graph.
           Does not remove atoms that no longer have any bonds as a result of
           this removal.
           """
           self._fingerprint = self._inchi = self._smiles = None
           return self.remove_edge(bond)


8 similar lines in 2 files
==fragment.py:[309:318]
==molecule.py:[1115:1124]
       def remove_atom(self, atom):
           """
           Remove `atom` and all bonds associated with it from the graph. Does
           not remove atoms that no longer have any bonds as a result of this
           removal.
           """
           self._fingerprint = self._inchi = self._smiles = None
           return self.remove_vertex(atom)


8 similar lines in 2 files
==fragment.py:[244:253]
==molecule.py:[1433:1442]
           return alist

       def get_all_labeled_atoms(self):
           """
           Return the labeled atoms as a ``dict`` with the keys being the labels
           and the values the atoms themselves. If two or more atoms have the
           same label, the value is converted to a list of these atoms.
           """
           labeled = {}

7 similar lines in 2 files
==fragment.py:[1048:1056]
==molecule.py:[2088:2096]
           """
           if self.symmetry_number == -1:
               self.calculate_symmetry_number()
           return self.symmetry_number

       def calculate_symmetry_number(self):
           """
           Return the symmetry number for the structure. The symmetry number

7 similar lines in 2 files
==fragment.py:[640:648]
==molecule.py:[1598:1606]
                           return True
                   else:
                       return False
               else:
                   if not self.is_mapping_valid(other, initial_map, equivalent=False):
                       return False

           # Do the isomorphism comparison

7 similar lines in 2 files
==fragment.py:[344:352]
==molecule.py:[1067:1075]
       def add_bond(self, bond):
           """
           Add a `bond` to the graph as an edge connecting the two atoms `atom1`
           and `atom2`.
           """
           self._fingerprint = self._inchi = self._smiles = None
           return self.add_edge(bond)


7 similar lines in 2 files
==fragment.py:[199:206]
==molecule.py:[1265:1272]
       def copy(self, deep=False):
           """
           Create a copy of the current graph. If `deep` is ``True``, a deep copy
           is made: copies of the vertices and edges are used in the new graph.
           If `deep` is ``False`` or not specified, a shallow copy is made: the
           original vertices and edges are used in the new graph.
           """

6 similar lines in 2 files
==fragment.py:[1579:1588]
==molecule.py:[2151:2160]
           if aromatic_rings is None:
               aromatic_rings = self.get_aromatic_rings(save_order=save_order)[0]

           total = self.get_radical_count()
           aromatic_atoms = set([atom for atom in itertools.chain.from_iterable(aromatic_rings)])
           aryl = sum([atom.radical_electrons for atom in aromatic_atoms])

           return total == aryl


6 similar lines in 2 files
==fragment.py:[1383:1391]
==molecule.py:[2554:2562]
           Uses entire range of cython's integer values to reduce chance of duplicates
           """

           global atom_id_counter

           for atom in self.atoms:
               atom.id = atom_id_counter
               atom_id_counter += 1

6 similar lines in 2 files
==fragment.py:[1347:1355]
==molecule.py:[2392:2400]
                                   aromatic_bonds_in_ring.append(self.get_bond(atom1, atom2))
                   else:  # didn't break so all atoms are carbon
                       if len(aromatic_bonds_in_ring) == 6:
                           aromatic_rings.append(ring0)
                           aromatic_bonds.append(aromatic_bonds_in_ring)

               return aromatic_rings, aromatic_bonds


6 similar lines in 2 files
==fragment.py:[1287:1295]
==molecule.py:[1723:1730]
           Skips the first line (assuming it's a label) unless `withLabel` is
           ``False``.
           """
           from rmgpy.molecule.adjlist import from_adjacency_list

           self.vertices, self.multiplicity = from_adjacency_list(adjlist, group=False, saturate_h=saturate_h)
           self.update_atomtypes(raise_exception=raise_atomtype_exception)

5 similar lines in 2 files
==fragment.py:[1538:1545]
==molecule.py:[1995:2002]
           else:
               # didn't fail
               return True

           # not returned yet? must be nonlinear
           return False


5 similar lines in 2 files
==fragment.py:[1483:1490]
==molecule.py:[1463:1470]
               if key in element_count:
                   element_count[key] += 1
               else:
                   element_count[key] = 1

           return element_count


5 similar lines in 2 files
==fragment.py:[153:158]
==molecule.py:[932:937]
                               'using InChI and ignoring SMILES.')
           if inchi:
               self.from_inchi(inchi)
               self._inchi = inchi
           elif smiles:
TOTAL lines=4840 duplicates=299 percent=6.18

</details>

**Comment by JacksonBurns** [comment]
An overall note for `fragment.py` - `assign_representative_molecule` should be called only once in the constructor and then _no where else_. Right now it is called every time it is needed, which regenerates the exact same molecule over and over. Huge performance cost.

Updates on the inheritance scheme follow.

The following methods can be directly inherited from `Molecule`, since they are currently directly copied from it anyway:
- [ ] add_atom
- [ ] add_bond
- [ ] assign_atom_ids
- [ ] atom_ids_valid
- [ ] clear_labeled_atoms
- [ ] contains_labeled_atom
- [ ] count_internal_rotors
- [ ] draw
- [ ] from_inchi
- [ ] generate_resonance_structures
- [ ] get_all_labeled_atoms
- [ ] get_bond
- [ ] get_charge_span
- [ ] get_labeled_atoms
- [ ] get_net_charge
- [ ] get_symmetry_number
- [ ] get_url
- [ ] has_bond
- [ ] is_atom_in_cycle
- [ ] is_bond_in_cycle
- [ ] is_identical
- [ ] is_isomorphic
- [ ] is_linear
- [ ] kekulize
- [ ] remove_atom
- [ ] remove_bond
- [ ] to_adjacency_list
- [ ] update_multiplicity

These methods must override the corresponding method in `Molecule` but are at least well-defined for `Fragment`:
- [ ] saturate_radical
- [ ] split
- [ ] to_smiles
- [ ] to_rdkit_mol
- [ ] update
- [ ] update_atomtypes
- [ ] update_lone_pairs
- [ ] from_adjacency_list
- [ ] get_aromatic_rings
- [ ] get_element_count
- [ ] get_formula
- [ ] get_molecular_weight
- [ ] get_n_atoms
- [ ] get_radical_count
- [ ] get_singlet_carbene_count
- [ ] is_aormatic
- [ ] is_aryl_radical
- [ ] is_radical
- [ ] is_subgraph_isomorphic
- [ ] is_surface_site
- [ ] merge
- [ ] calculate_cp0
- [ ] calculate_cpinf
- [ ] calculate_symmetry_number
- [ ] contains_surface_site
- [ ] copy

These methods can be left in fragment likely as they are without any problems, since the extend the `Molecule` class:
- [ ] sliceitup_aliph
- [ ] sliceitup_arom
- [ ] replace_cutting_label
- [ ] pattern_call
- [ ] get_representative_molecule
- [ ] from_rdkit_mol
- [ ] from_smiles_like_string
- [ ] cut_molecule
- [ ] detect_cutting_label
- [ ] assign_representative_molecule
- [ ] assign_representative_species
- [ ] check_in_ring

Lastly, these methods are present in `Molecule` and may or may not be well-defined for `Fragment` (will likely need to be overridden and raise a `TypeError` if reasonable behavior for `Fragment` cannot be determined):
- [ ] connect_the_dots
- [ ] delete_hydrogens
- [ ] enumerate_bonds
- [ ] find_h_bonds
- [ ] find_isomorphism
- [ ] find_subgraph_isomorphisms
- [ ] from_augmented_inchi
- [ ] from_smarts
- [ ] from_smiles
- [ ] from_xyz
- [ ] generate_h_bonded_structures
- [ ] get_adatoms
- [ ] get_bonds
- [ ] get_desorbed_molecules
- [ ] get_deterministic_sssr
- [ ] get_nth_neighbor
- [ ] get_radical_atoms
- [ ] get_surface_sites
- [ ] has_atom
- [ ] has_charge
- [ ] has_halogen
- [ ] has_lone_pair
- [ ] is_heterocyclic
- [ ] remove_h_bonds
- [ ] replace_halogen_with_hydrogen
- [ ] remove_van_der_waals_bonds
- [ ] saturate_unfilled_valence
- [ ] sort_atoms
- [ ] to_augmented_inchi
- [ ] to_augmented_inchi_key
- [ ] to_group
- [ ] to_inchi
- [ ] to_inchi_key
- [ ] to_single_bonds
- [ ] to_smarts
- [ ] count_aromatic_rings

These will be checked of as they are fixed/implemented/validated.

**Comment by JacksonBurns** [comment]
This PR is ready for review. It does not add or remove any features, only significantly simplifies the implementation of `Fragment` by making it inherit from `Molecule`.

I have done all the reformatting in the first commit, so actual substantial code changes can be seen individually in the second two commits. It's all in one file anyway, so hopefully not too much trouble.

None of the 'possibly ill-defined for fragment' methods that I discussed [here](https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2413#issuecomment-1516977217) have actually been an issue in the CI, though this might be due to them not being covered. #2400 mentioned agreed that this inheritance scheme is possibly a good idea though it depends on the finer details of `Fragment`.

I am marking as ready for review with this limitation in mind. If we want to tease out all the possible issues with 'possibly ill-defined' methods, I think should go into main and we should get users on it.

Requesting reviews from @rwest and @alongd (from #2400) and also requesting @lily90502 to run a Fragment model using RMG from this branch to see if any errors pop up.

**Comment by JacksonBurns** [comment]
Latest changes make `Fragment` compatible with `RMS` and add a simple regression test to check for the functionality of AFM in the future. Will let CI run but this should be close now.

**Comment by JacksonBurns** [comment]
Additional note - as mentioned in the extended description of 3644fbf this PR now includes a patch that is required to make `Fragment` work (i.e. the current release and main both do **not** have a functional fragment feature).

**Comment by JacksonBurns** [comment]
I am re-requesting reviews since this PR is now set to fix bugs that are affecting the currently deployed version of RMG (see https://github.com/ReactionMechanismGenerator/RMG-Py/issues/2548).

Please let me know what fixes are needed, @lily90502 @rwest and @hwpang

**Comment by JacksonBurns** [comment]
This PR should be a twin PR with RMS#216 which introduces some compatibility fixes with Fragment inside of RMS. The CI should be switched to use that branch for that PR as well to test the compatibility fixes.

File: test/regression/fragment/input.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
+database(
+    thermoLibraries=["primaryThermoLibrary", "SulfurLibrary", "DFT_QCI_thermo"],
```
Comment (hwpang) [inline test/regression/fragment/input.py:2]: Probably a nitpick but since you are only generating for all carbon fragment, the sulfur library is not needed

File: rmgpy/molecule/fragment.py [hunk 2]
```diff
@@ -32,32 +32,38 @@
...
+        # define a new custom atom type
+        self.atomtype = AtomType(label=label, generic=[], specific=[])
+        # accessed by other methods in Fragment
+        self.name = name
         self.isotope = -1
-        self.id = id
-        self.mass = 0
-        self.site = ""
-        self.morphology = ""
-
```
Comment (hwpang) [inline rmgpy/molecule/fragment.py]: Why are these added in previous commit but deleted here?

File: rmgpy/rmg/reactors.py [hunk 3]
```diff
@@ -690,6 +691,28 @@ def to_rms(obj, species_names=None, rms_species_list=None, rmg_species=None):
...
+                bondnum=bondnum,
+                diffusion=rms.EmptyDiffusivity(),
+                radius=0.0,
+                radicalelectrons=obj.molecule[0].multiplicity - 1,
+                molecularweight=0.0,
+                henrylawconstant=rms.EmptyHenryLawConstant(),
+                liquidvolumetricmasstransfercoefficient=rms.EmptyLiquidVolumetricMassTransferCoefficient(),
+                comment=obj.thermo.comment,
+            )
+
```
Comment (hwpang) [inline rmgpy/rmg/reactors.py:715]: I don't think we should hard code that fragment does not have diffusion, have no molecular weights, and have no kLA and kH. We should reuse what we have below for molecule, but change the following parts
- estimates of the species radius for fragments should be done with the representative molecule
- use this estimate of radius from the representative molecule to estimate stokes diffusivity, instead of hard coded no diffusion limitation
- same goes for kLA and kH for fragments. Fragments' solvation free energy should be estimated using their representative molecule.

**Review by hwpang — CHANGES_REQUESTED** [review]
Thanks for the PR. I had some request of changes.

File: rmgpy/rmg/reactors.py (hunk 3 above)
Comment (hwpang) [reply rmgpy/rmg/reactors.py:715]: I can push some changes regarding this because this falls into what I have been helping Yen-Ting with over the past year if you're fine with it

Comment (JacksonBurns) [reply rmgpy/rmg/reactors.py:715]: Please do, thank you.

File: rmgpy/molecule/fragment.py [hunk 4]
```diff
@@ -32,32 +32,38 @@
...
+            charge=0,
+            label=label,
+            lone_pairs=0,
+            site="",
+            coords=None,
+            id=id,
+            props=None,
+        )
+        # define a new custom atom type
+        self.atomtype = AtomType(label=label, generic=[], specific=[])
```
[the remark this answers is not in the document]
Comment (hwpang) [reply rmgpy/molecule/fragment.py:62]: I can also push something to this if you're ok with it

File: test/regression/fragment/input.py (hunk 1 above)
Comment (JacksonBurns) [reply test/regression/fragment/input.py:2]: Good catch, done

File: rmgpy/molecule/adjlist.py [hunk 5]
```diff
@@ -91,7 +91,7 @@ def check_partial_charge(atom):
         the theoretical one:
 
         """
-        if atom.symbol == 'X':
+        if atom.symbol in {'X','L','R'}:
```
[the remark this answers is not in the document]
Comment (JacksonBurns) [reply rmgpy/molecule/adjlist.py:95]: Actually, `update_atomtypes` in fragment skips cuttinglabels, so I think this is safe, albeit probably grounds for confusion in the future.

File: rmgpy/molecule/fragment.py (hunk 2 above)
Comment (JacksonBurns) [reply rmgpy/molecule/fragment.py]: I changed inheritance schemes a couple times, will clean up commit history closer to merging.

**Comment by JacksonBurns** [comment]
@hwpang I have made the simpler of the requested edits, could you push your changes as well so we can see where this PR is at?

**Review by hwpang — APPROVED** [review]
Looks good to me.
